docs/features/pouch_with_plugin.md

```diff
@@ -1,148 +1,161 @@
+	// used to sort the pre start array that pass to runc, network plugin hook always has priority value 0.
+	PreStart(interface{}) ([]int, [][]string, error)
+
+	// PreCreateEndpoint accepts the container id and env of this container and returns the priority of this endpoint
+	// and if this endpoint should enable resolver and a map which will be used as generic params to create endpoints of
+	// this container
+	PreCreateEndpoint(string, []string) (priority int, disableResolver bool, genericParam map[string]interface{})
+
+	// PreUpdate defines plugin point where receives a container update request, in this plugin point user
+	// could change the container update body passed-in by http request body
+	PreUpdate(io.ReadCloser) (io.ReadCloser, error)
+
+	// PostUpdate called after update method successful,
+	// the method accepts the rootfs path and envs of container
+	PostUpdate(string, []string) error
 }
-
 ```
 
-These two Plugin symbols will be fetch by name `DaemonPlugin` and `ContainerPlugin` from shared object file like this:
+### daemon plugin
+
+* pre-start daemon point, at this point you can start assist processes like network plugins and dfget proxy which need by pouchd and whose life cycle is the same as pouchd.
+* pre-stop daemon point, at this point you can stop the assist processes gracefully, but the trigger of this point is not a promise, because pouchd may be killed by SIGKILL.
+
+Defined as follow:
 
 ```
-p, _ := plugin.Open("path_to_shared_object_file")
-daemonPlugin, _ := p.Lookup("DaemonPlugin")
-containerPlugin, _ := p.Lookup("ContainerPlugin")
+// DaemonPlugin defines places where a plugin will be triggered in pouchd lifecycle
+type DaemonPlugin interface {
+	// PreStartHook is invoked by pouch daemon before real start, in this hook user could start dfget proxy or other
+	// standalone process plugins
+	PreStartHook() error
+
+	// PreStopHook is invoked by pouch daemon before daemon process exit, not a promise if daemon is killed, in this
+	// hook user could stop the process or plugin started by PreStartHook
+	PreStopHook() error
+}
 ```
 
-## example
+### volume plugin
 
-define two plugin symbols which only print some logs at correspond point:
+* pre-volume-create volume point, at this point you can change the create volume's config as you want, add your default volume's options.
 
+Defined as follow:
+
+```
+// VolumePlugin defines places where a plugin will be triggered in volume lifecycle
+type VolumePlugin interface {
+	// PreVolumeCreate defines plugin point where receives an volume create request, in this plugin point user
+	// could change the volume create body passed-in by http request body
+	PreVolumeCreate(*types.VolumeCreateConfig) error
+}
 ```
-package main
 
-import (
-    "fmt"
-    "io"
-)
+### cri plugin
 
-var ContainerPlugin ContPlugin
+* pre-create-container cri container point, at this point you can update the container config what it will be created, such as update the container's envs or labels.
```

Comment:
```suggestion
* pre-create-container cri point, at this point you can update the container config to what it will be created, such as update the container's envs or labels.
```